tests/ui/redundant_async_block.rs

```diff
@@ -0,0 +1,31 @@
+// run-rustfix
+
+#![allow(unused)]
+#![warn(clippy::redundant_async_block)]
+
+async fn foo() -> u32 {
+    42
+}
+
+async fn bar() -> String {
+    let s = String::from("some string");
+    let f = async { s.deref().to_owned() };
+    let x = async { f.await };
+    x.await
+}
+
+fn main() {
+    let fut1 = async { 42 };
+    let fut2 = async { fut1.await };
+
+    let fut1 = async { 42 };
+    let fut2 = async move { fut1.await };
+
+    let fut = async { async { 42 }.await };
```

Comment:
I'd like to see a test where the future is macro-generated, a test where the whole block *outside* the future / with the future is macro-generated by an crate-internal/external macro each.

The internal ones may lint, but with the correct snippet; the external ones I think should not lint (because we don't expect people to be able to change code external to their crate). I want to be sure the `span.from_expansion()` check is sufficient.

Reply:
Checking inside a macro needs a bit more work, because the awaiten future must not come from a macro parameter, as it might contain `.await` itself.

Reply:
Right now, it doesn't look into macros at all. I've added a `MISSED OPPORTUNITY` marker, but I'm not sure how to check inside the macro that no parameter (or another macro) is used as the `.await` receiver.

Reply:
That's all right. :+1: